sql: hoist FunctionProperties to the built-in function name, instead of individual overloads

knz opened the pull request:
tldr: this patchset hoists the properties of built-in functions next to the name, instead of the level of individual overload implementation as previously.

This is a precursor to the fix to  #20511.

Fixes #26354.

cc @andy-kimball

knz committed:
sql: simplify (*FuncExpr).TypeCheck()

Release note: None

knz committed:
sql: define common FunctionProperties for built-in functions

This patch introduces a new set of `FunctionProperties` common to all
overloads of a built-in function. This is meant to simplify the
derivation of properties from `FuncExprs`, so that properties can be
derived without performing overload resolution.

This patch currently embeds the new `FunctionProperties` in each
overload still, so that a successful build with this patch forms an
assertion that all current overload definitions indeed define the same
properties for a given built-in.

(I checked that the assertion indeed holds.)

A future patch may hoist the common properties to a single struct
shared by all the overloads.

Release note: None

knz committed:
sql/sem: rename Builtin to OverloadDefinition

Release note: None

knz committed:
sql/sem: hoist the function properties to the overload group

Prior to this patch, CockroachDB would associate function properties
like "this function is impure" or "this function is privileged" to
individual overloads, so that two variants of the same function name
could have diverging properties.

This makes the code unnecessarily complex and also flies in the face
of common sense.

This patch cleans up the situation by hoisting the properties to the
level of the overload group.

Some minor "cosmetic" fixes in the process:

- the generated function documentation is now better categorized
- this includes the output of `\hf` in the CLI shell
- `current_date()` now properly marked as impure

Also various objects from the `builtins` package are now unexported
(they were never meant to be exported in the first place). An accessor
function remains (`GetBuiltinProperties`) but really other users
should use `tree.FunctionDefinition` instead.

Release note (bug fix): Fixed a bug where a prepared query would not
produce the right value for `current_date()` if prepared on one day
and executed on the next.

knz committed:
sql: minor cleanup to (*FuncExpr).TypeCheck()

The check for privileged functions can now be hoisted to before
argument analysis. Do it.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26370)
<!-- Reviewable:end -->

knz committed:
docs: update the codelab for SQL built-in functions

Release note: None

jordanlewis commented:
Generally <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> although I didn't carefully review the whole patch on account of this mainly being a refactor.

---

Review status: 2 of 31 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[docs/generated/sql/functions.md, line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26370#-LEBX6vtBiTeR9DF6WjI:-LEBX6vtBiTeR9DF6WjJ:b-rz9zoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/70fbdd47353245c3f1cf98ec0aac9b95c03165f7/docs/generated/sql/functions.md#L1)):*
> ```Markdown
> ### ANYELEMENT Functions
> ```

This seems to have been wholesale deleted. was that intended?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26370#-:-LEBXzAv24jysbAGjIa2:b-y021bv)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
One naming nit is that  you could just call it Overload or OverloadDef instead of OverloadDefinition since we  have to type that out a lot.

---

Review status: 2 of 31 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26370#-:-LEBYGhW8S3aCnle-z7o:bkxsx5s)*
<!-- Sent from Reviewable.io -->

andy-kimball commented on pkg/sql/opt/memo/private_defs.go:
```diff
@@ -31,9 +31,10 @@ type PrivateID uint32
-	Overload *tree.Builtin
+	Name       string
+	Type       types.T
+	Properties *tree.FunctionProperties
```
This feels clunky to me. I should be able to have a single pointer that holds onto a function definition, including its properties and its overloads. Why can't we have a `BuiltinDef` struct, like this:
```
type BuiltinDef struct {
  Props BuiltinProps
  Overloads []BuiltinOverload
}
```
This way, I can easily get a pointer to whatever I need - props, all overloads, any one overload, or the entire definition.

andy-kimball commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -94,53 +96,83 @@ func categorizeType(t types.T) string {
+
 // Builtins contains the built-in functions indexed by name.
-var Builtins = map[string][]tree.Builtin{
+var Builtins = map[string]BuiltinDefinition{
```
Can you make this a `*BuiltinDefinition` instead? Otherwise it forces the code that accesses the map to make a copy of the struct, since you can't get a pointer to a map value.

knz commented:
Review status: 2 of 31 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/generated/sql/functions.md, line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26370#-LEBX6vtBiTeR9DF6WjI:-LEDtp328mKKZMOKAT2P:bfnzw7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/70fbdd47353245c3f1cf98ec0aac9b95c03165f7/docs/generated/sql/functions.md#L1)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This seems to have been wholesale deleted. was that intended?
</blockquote></details>

The functions have be re-categorized in "JSON" where they belong.

---

*[pkg/sql/opt/memo/private_defs.go, line 36 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26370#-LEBAJMmYTpdgwUNmJwf-r6-36:-LEDtwfO5POFtoqVIg7-:bvhnzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/06c32bc0982dcc68f79cfa083cc24169668a9f79/pkg/sql/opt/memo/private_defs.go#L36)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This feels clunky to me. I should be able to have a single pointer that holds onto a function definition, including its properties and its overloads. Why can't we have a `BuiltinDef` struct, like this:
```
type BuiltinDef struct {
  Props BuiltinProps
  Overloads []BuiltinOverload
}
```
This way, I can easily get a pointer to whatever I need - props, all overloads, any one overload, or the entire definition.
</blockquote></details>

AFAIK `FuncOpDef` needs a particular overload, not the full set.
If this structhad pointed to a `*tree.FunctionDefinition` (which exists already, and does what you say you want), I'd not have needed change this.

If you want everything, change the code to use `tree.FunctionDefinition` as needed.

---

*[pkg/sql/sem/builtins/builtins.go, line 129 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26370#-LEBAJMmYTpdgwUNmJwV-r6-129:-LEDuF3l1JLAtfNhbjMR:b-5b9sjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/06c32bc0982dcc68f79cfa083cc24169668a9f79/pkg/sql/sem/builtins/builtins.go#L129)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can you make this a `*BuiltinDefinition` instead? Otherwise it forces the code that accesses the map to make a copy of the struct, since you can't get a pointer to a map value.
</blockquote></details>

`BuiltinDefinition` is not meant for public consumption (it's only exported to generate help strings from a different package). Use `FunctionDefinition` instead.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26370)*
<!-- Sent from Reviewable.io -->

knz committed:
sql/sem: move AsJSON to package tree where it belongs

Release note: None

knz committed:
sql/sem: rename `OverloadDefinition` to simply `Overload`

.. because "we have to type that a lot".

Requested by @jordanlewis.

Release note: None

knz commented:
Renamed `OverloadDefinition` to `Overload` as requested.

knz committed:
sql/sem: add checks that impure functions do not get constant-folded

These tests were missing.

Release note: None

knz commented:
bors r+

andy-kimball commented:
*[pkg/sql/opt/memo/private_defs.go, line 36 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26370#-LEBAJMmYTpdgwUNmJwf-r6-36:-LEFNY4n89bseMAx_jft:b-tk1b2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/06c32bc0982dcc68f79cfa083cc24169668a9f79/pkg/sql/opt/memo/private_defs.go#L36)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

AFAIK `FuncOpDef` needs a particular overload, not the full set.
If this structhad pointed to a `*tree.FunctionDefinition` (which exists already, and does what you say you want), I'd not have needed change this.

If you want everything, change the code to use `tree.FunctionDefinition` as needed.
</blockquote></details>

Yes, you're correct, we do need the overload. Perhaps I'll store the `FunctionDefinition` + int index of overload. Then I can also get rid of the `Name` field, since that's already stored in the definition as well.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26370)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 39 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/builtins/builtins.go, line 129 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26370#-LEBAJMmYTpdgwUNmJwV-r6-129:-LEFR6Xx2AK9g406qi7v:bsyw1a7) ([raw file](https://github.com/cockroachdb/cockroach/blob/06c32bc0982dcc68f79cfa083cc24169668a9f79/pkg/sql/sem/builtins/builtins.go#L129)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

OK, looks like you found a way to not export it, which clarifies its intended use. How come you can't just  use `FunctionDefinition` directly here? Looking at the latest code, it appears that we're storing almost the same thing, first as `builtinDefinition` and then as `FunctionDefinition`, and copying one to the other.
</blockquote></details>

No need to add the `Name` field in each entry of the `builtins` map since the map key already has it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26370)*
<!-- Sent from Reviewable.io -->
